### PR TITLE
vimPlugins.kulala-nvim: init at 2024-09-30

### LINKED ISSUE #325970: Package request: vimPlugins-kulala.nvim
**Project description**

A minimal REST-Client Interface for Neovim.

**Metadata**

* homepage URL: https://github.com/mistweaverco/kulala.nvim
* source URL: https://github.com/mistweaverco/kulala.nvim
* license: mit
* platforms: all

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by CnTeng**
## Description of changes
https://github.com/mistweaverco/kulala.nvim
close https://github.com/NixOS/nixpkgs/issues/325970

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by PerchunPak** [comment]
This plugin depends on curl, it is probably a good idea to patch it like here: https://github.com/NixOS/nixpkgs/blob/a4638fa3cbec7bdd5a79b212a75f1dc00957e011/pkgs/applications/editors/vim/plugins/overrides.nix#L316-L323
Lines, where curl is used:
https://github.com/mistweaverco/kulala.nvim/blob/7649576ca459146658b4c415d1a320bbedc3f564/lua/kulala/graphql/init.lua#L21
https://github.com/mistweaverco/kulala.nvim/blob/7649576ca459146658b4c415d1a320bbedc3f564/lua/kulala/parser/init.lua#L448

**Review by jlesquembre — APPROVED** [review]

**Review by jlesquembre — CHANGES_REQUESTED** [review]
You forgot to update https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/editors/vim/plugins/vim-plugin-names.
Other than that, LGTM.

**Review by PerchunPak — CHANGES_REQUESTED** [review]
Don't know why, but it fails to run with this error:

```
Failed to run `config` for kulala.nvim

...d-0.10.1/share/nvim/runtime/lua/vim/treesitter/query.lua:252: Query error at 1:2. Invalid node type "section":
(section (request) @request) @section
 ^


# stacktrace:
  - /nix/store/g2a2ipz91mj1pqqca063idj21sg47wbd-neovim-unwrapped-0.10.1/share/nvim/runtime/lua/vim/treesitter/query.lua:252 _in_ **fn**
  - /nix/store/g2a2ipz91mj1pqqca063idj21sg47wbd-neovim-unwrapped-0.10.1/share/nvim/runtime/lua/vim/func/_memoize.lua:58 _in_ **parse**
  - /nix/store/z98vs4iz6hixp4dg4zmxv5ajzvavzcac-vim-pack-dir/pack/myNeovimPackages/start/kulala.nvim/lua/kulala/parser/treesitter.lua:12
  - /nix/store/z98vs4iz6hixp4dg4zmxv5ajzvavzcac-vim-pack-dir/pack/myNeovimPackages/start/kulala.nvim/lua/kulala/parser/init.lua:11
  - /nix/store/z98vs4iz6hixp4dg4zmxv5ajzvavzcac-vim-pack-dir/pack/myNeovimPackages/start/kulala.nvim/lua/kulala/ui/init.lua:6
  - /nix/store/z98vs4iz6hixp4dg4zmxv5ajzvavzcac-vim-pack-dir/pack/myNeovimPackages/start/kulala.nvim/lua/kulala/init.lua:1
  - /nix/store/skk4qac1ds1957i9dpq2zcin80bcnra4-nixCats-special-rtp-entry-LuaConfig/lua/nixCatsUtils/lazyCat.lua:135 _in_ **setup**
  - /nix/store/skk4qac1ds1957i9dpq2zcin80bcnra4-nixCats-special-rtp-entry-LuaConfig/lua/lazy-plugins.lua:3
  - /nix/store/skk4qac1ds1957i9dpq2zcin80bcnra4-nixCats-special-rtp-entry-LuaConfig/init.lua:51
  - /nix/store/xysvz9j297w2m754hljznh9ynhxpvhn1-init.lua:19
```

Also, this plugin now depends on treesitter HTTP grammar as well: https://kulala.mwco.app/docs/getting-started/requirements/#syntax-highlighting (crashes without it)

**Comment by CnTeng** [comment]
But, there's no way to add treesitter-http   into plugin dependency. At least, I don't know how to.

**Comment by PerchunPak** [comment]
There is:

```nix
  kulala-nvim = super.kulala-nvim.overrideAttrs {
    dependencies = with self; [ (nvim-treesitter.withPlugins (p: [ p.http ])) ];
  };
```

https://github.com/NixOS/nixpkgs/blob/fab8ea932f8f896fb8eaf882ec154779c4e8171e/pkgs/applications/editors/vim/plugins/nvim-treesitter/overrides.nix#L33-L37

**Comment by jlesquembre** [comment]
Not sure if this is the best approach, but `image-nvim` does this:
https://github.com/NixOS/nixpkgs/blob/fab8ea932f8f896fb8eaf882ec154779c4e8171e/pkgs/applications/editors/vim/plugins/overrides.nix#L899-L904

**Comment by PerchunPak** [comment]
Yeah, it looks like a legacy approach. I think, better to use `withPlugins`

**Comment by jlesquembre** [comment]
```nix
  kulala-nvim = super.kulala-nvim.overrideAttrs {
    dependencies = with self; [ (nvim-treesitter.withPlugins (p: [ p.http ])) ];
  };
```
The problem with that approach is that it seems to override the user configuration. For me, I have `nvim-treesitter.withAllGrammars` set, but it gets overridden, disabling all other TS grammars.

I personally would merge the PR as it is.

**Review by jlesquembre — APPROVED** [review]

**Comment by CnTeng** [comment]
> Don't know why, but it fails to run with this error:
> 
> ```
> Failed to run `config` for kulala.nvim
> 
> ...d-0.10.1/share/nvim/runtime/lua/vim/treesitter/query.lua:252: Query error at 1:2. Invalid node type "section":
> (section (request) @request) @section
>  ^
> 
> 
> # stacktrace:
>   - /nix/store/g2a2ipz91mj1pqqca063idj21sg47wbd-neovim-unwrapped-0.10.1/share/nvim/runtime/lua/vim/treesitter/query.lua:252 _in_ **fn**
>   - /nix/store/g2a2ipz91mj1pqqca063idj21sg47wbd-neovim-unwrapped-0.10.1/share/nvim/runtime/lua/vim/func/_memoize.lua:58 _in_ **parse**
>   - /nix/store/z98vs4iz6hixp4dg4zmxv5ajzvavzcac-vim-pack-dir/pack/myNeovimPackages/start/kulala.nvim/lua/kulala/parser/treesitter.lua:12
>   - /nix/store/z98vs4iz6hixp4dg4zmxv5ajzvavzcac-vim-pack-dir/pack/myNeovimPackages/start/kulala.nvim/lua/kulala/parser/init.lua:11
>   - /nix/store/z98vs4iz6hixp4dg4zmxv5ajzvavzcac-vim-pack-dir/pack/myNeovimPackages/start/kulala.nvim/lua/kulala/ui/init.lua:6
>   - /nix/store/z98vs4iz6hixp4dg4zmxv5ajzvavzcac-vim-pack-dir/pack/myNeovimPackages/start/kulala.nvim/lua/kulala/init.lua:1
>   - /nix/store/skk4qac1ds1957i9dpq2zcin80bcnra4-nixCats-special-rtp-entry-LuaConfig/lua/nixCatsUtils/lazyCat.lua:135 _in_ **setup**
>   - /nix/store/skk4qac1ds1957i9dpq2zcin80bcnra4-nixCats-special-rtp-entry-LuaConfig/lua/lazy-plugins.lua:3
>   - /nix/store/skk4qac1ds1957i9dpq2zcin80bcnra4-nixCats-special-rtp-entry-LuaConfig/init.lua:51
>   - /nix/store/xysvz9j297w2m754hljznh9ynhxpvhn1-init.lua:19
> ```
> 
> Also, this plugin now depends on treesitter HTTP grammar as well: https://kulala.mwco.app/docs/getting-started/requirements/#syntax-highlighting (crashes without it)

I think the http treesitter in nixpkgs is old or kulala-nvim has problem.

**Comment by PerchunPak** [comment]
> I think the http treesitter in nixpkgs is old or kulala-nvim has problem

Looks like a kulala problem. Though the plugin is in active development and enough time passed, I think if you update the PR everything will be fixed

> And adding treesitter into plugin dependencies is not a good solution.

Specifing `nvim-treesitter-parsers.http` in dependencies doesn't overwrite other grammars, so it is safe ig

**Comment by CnTeng** [comment]
The error is due to https://github.com/mistweaverco/kulala.nvim/issues/213

And, I think everyone using Treesitter has installed it individually.

**Comment by PerchunPak** [comment]
`withPlugins (p: [ p.http ])` just returns `nvim-treesitter` with `nvim-treesitter-parsers.http` in `nvim-treesitter.dependencies`, so just add `nvim-treesitter-parsers.http` and treesitter in dependencies and it will be safe

**Comment by gorillamoe** [comment]
Treesitter should load just on demand for now (if you're enabled treesitter= true via opts).

So the problem should be fixed.

**Comment by PerchunPak** [comment]
Well, it crashed for me when I didn't have HTTP grammar installed

**Comment by gorillamoe** [comment]
> Well, it crashed for me when I didn't have HTTP grammar installed

Will check, shouldn't crash, because it should be completely optional at the current state.

**Comment by gorillamoe** [comment]
Have checked with the latest commit on main, ts http grammar not installed. Won't crash, just works. 🤷🏾

**Comment by gorillamoe** [comment]
You need to change the replacement of curl when using the latest version. It's configurable via opts now.

https://github.com/mistweaverco/kulala.nvim/blob/f6bbeeff6f8eaa2dc9b09148b6e0746ced5951ca/lua/kulala/config/init.lua#L8

**Comment by CnTeng** [comment]
Ok, I will do it later.

**Comment by CnTeng** [comment]
Now, treesitter-http is optional, so there's no need to add it to the dependencies.

**Review by PerchunPak — APPROVED** [review]
Verified that it works without treesitter

CC @GaetanLepage

File: pkgs/applications/editors/vim/plugins/overrides.nix [hunk 1]
```diff
@@ -911,6 +911,14 @@
     passthru.python3Dependencies = ps: [ ps.jupytext ];
   };
 
+  kulala-nvim = super.kulala-nvim.overrideAttrs {
+    buildInputs = [ curl ];
+    postPatch = ''
+      substituteInPlace lua/kulala/config/init.lua \
+        --replace 'curl_path = "curl"' 'curl_path = "${lib.getExe curl}"'
+    '';
```
Comment (GaetanLepage) [inline pkgs/applications/editors/vim/plugins/overrides.nix:919]: Isn't there the new "passtrhu/snippet" for declaring those opt-in patches ?
Or maybe it doesn't apply here.

Comment (PerchunPak) [reply pkgs/applications/editors/vim/plugins/overrides.nix:919]: There is (https://github.com/NixOS/nixpkgs/pull/334913), but does it apply to `.setup` functions?

https://github.com/mistweaverco/kulala.nvim/blob/f6bbeeff6f8eaa2dc9b09148b6e0746ced5951ca/lua/kulala/config/init.lua#L93

Also, it is hard to find, so I would expect false bug reports about it

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 332758` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>vimPlugins.kulala-nvim</li>
    <li>vimPluginsUpdater</li>
  </ul>
</details>

File: pkgs/applications/editors/vim/plugins/overrides.nix (hunk 1 above)
Comment (CnTeng) [reply pkgs/applications/editors/vim/plugins/overrides.nix:919]: It is hard to add passthru here, it config like
```lua
require("kulala").setup({
})
```

Comment (CnTeng) [reply pkgs/applications/editors/vim/plugins/overrides.nix:919]: Someone use other way manager neovim plugins, just use pkgs without other config. And, I think this passthru is useful to vim.g.xxx.
If curl path setted by vim.g.curl_path, passthru is the best way.

**Comment by gorillamoe** [comment]
We're planning on having treesitter-http-grammar as a dependecy soon, but won't break if it's not available. We will just notify the user via a message, that Kulala won't work without the grammar.

Is this something you guys and girls can live with, or do we (or you) need to update the pkg then?

See this discussion for further information: https://github.com/mistweaverco/kulala.nvim/pull/225

**Comment by CnTeng** [comment]
I add treesitter to dependencies. But I think is not the best way to solve treesitter dependencies.

I have no plan about adding initLua passthru. `require("xxx")` will load plugin, and this will cause problems if lazy or lz.n be used.

**Review by PerchunPak — CHANGES_REQUESTED** [review]
> We're planning on having treesitter-http-grammar as a dependecy soon

I still feel this plugin moves too fast for nixpkgs :)

File: pkgs/applications/editors/vim/plugins/overrides.nix [hunk 2]
```diff
@@ -911,6 +911,15 @@
     passthru.python3Dependencies = ps: [ ps.jupytext ];
   };
 
+  kulala-nvim = super.kulala-nvim.overrideAttrs {
+    dependencies = with self; [ (nvim-treesitter.withPlugins (p: [ p.http ])) ];
```
[the remark this answers is not in the document]
Comment (CnTeng) [reply pkgs/applications/editors/vim/plugins/overrides.nix:915]: What is the right way? I don't known.
https://github.com/NixOS/nixpkgs/blob/1355a0cbfeac61d785b7183c0caaec1f97361b43/pkgs/applications/editors/vim/plugins/overrides.nix#L1413
https://github.com/NixOS/nixpkgs/blob/1355a0cbfeac61d785b7183c0caaec1f97361b43/pkgs/applications/editors/vim/plugins/overrides.nix#L1448
https://github.com/NixOS/nixpkgs/blob/1355a0cbfeac61d785b7183c0caaec1f97361b43/pkgs/applications/editors/vim/plugins/overrides.nix#L1491

[the remark this answers is not in the document]
Comment (GaetanLepage) [reply pkgs/applications/editors/vim/plugins/overrides.nix:915]: @teto might know better

**Comment by gorillamoe** [comment]
What's the status here? Can I help you guys and girls with something to get this merged?

**Comment by PerchunPak** [comment]
After a discussion in Matrix, we decided that the best way is to add dependencies only to PATH in wrapper (maybe will be implemented in #344541).

For now, @CnTeng, please implement solution I provided in that review above. Unfortunately, because of the life situation, I don't really have time and cannot test right now if those plugins are broken.

**Review by PerchunPak — APPROVED** [review]
Finally LGTM

CC @GaetanLepage

**Review by GaetanLepage — APPROVED** [review]

**Pull request merged**